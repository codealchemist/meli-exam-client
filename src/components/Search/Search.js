import React from 'react'
import styled from 'styled-components'
import { withRouter } from 'react-router-dom'
import Colors from '../Colors'
import Icon from '../Icon'
import { search as searchApi } from '../../api'
import Context from '../../Context'

const Wrapper = styled.div`
  flex: 1;
  position: relative;

  input {
    height: 39px;
    box-shadow: 0 1px 2px 0 rgba(0, 0, 0, 0.2);
    padding: 7px 60px 9px 15px;
    border-radius: 2px;
    background-color: ${Colors.default.light1};
    border: 0 rgba(0, 0, 0, 0.2);
    color: ${Colors.default.dark1};
    font-size: 16px;
    width: 100%;
    margin: 0;
  }
`

class Search extends React.PureComponent {
  state = {
    query: null
  }

  search = context => {
    console.log('SEARCH context', context)
    searchApi(this.state.query).then(response => {
      console.log('SEARCH RESPONSE', response)
      context.setSearchResponse(response)

      // Redirect to Results Page.
      const { history } = this.props
      history.push(`/items?search=${this.state.query}`)
    })
  }

  onChange = (e, context) => this.setState({ query: e.target.value })

  onKey = ({ key }, context) => {
    if (key !== 'Enter') return
    this.search(context)
  }

  render() {
    return (
      <Context.Consumer>
        {context => (
          <Wrapper>
            <input
              type="search"
              placeholder="Buscar productos, marcas y mÃ¡s..."
              maxLength="120"
              autoFocus
              onChange={e => this.onChange(e, context)}
              onKeyPress={e => this.onKey(e, context)}
            />
            <Icon
              type="nav-icon-search"
              background={Colors.default.light2}
              onClick={() => this.search(context)}
            />
          </Wrapper>
        )}
      </Context.Consumer>
    )
  }
}

export default withRouter(Search)
