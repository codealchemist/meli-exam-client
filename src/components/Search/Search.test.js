import React from 'react'
import { mount } from 'enzyme'
import Search from './Search'

// Turn off console log.
global.console = { log: jest.fn() }

jest.mock('../Icon', () => 'div')
jest.mock('react-router-dom', () => ({
  withRouter: component => component
}))
jest.mock('../../Context', () => ({
  Consumer: obj =>
    obj.children({
      setSearchResponse: jest.fn()
    })
}))
jest.mock('../../api', () => ({
  search: () => ({
    then: callback => callback('METAL')
  })
}))

describe('Search', () => {
  it('should render', () => {
    const wrapper = mount(<Search />)
    expect(wrapper).toMatchSnapshot()
  })

  it('should update query', () => {
    const wrapper = mount(<Search />)
    const instance = wrapper.instance()
    instance.setState = jest.fn()
    const input = wrapper.find('input')
    input.simulate('change', { target: { value: 'Yeah!' } })
    expect(instance.setState).toHaveBeenCalledWith({ query: 'Yeah!' })
  })

  it('should search', () => {
    const history = {
      push: jest.fn()
    }
    const wrapper = mount(<Search history={history} />)
    const instance = wrapper.instance()
    instance.setState = jest.fn()
    instance.state.query = 'Yes sir.'

    const input = wrapper.find('input')
    input.simulate('change', { target: { value: 'Yeah!' } })
    input.simulate('keypress', { key: 'Enter' })
    expect(history.push).toHaveBeenCalledWith('/items?search=Yes sir.')
  })

  it('should set search response on context', () => {
    const history = {
      push: jest.fn()
    }
    const wrapper = mount(<Search history={history} />)
    const instance = wrapper.instance()
    instance.state.query = 'Dude'

    const context = {
      setSearchResponse: jest.fn()
    }
    instance.search(context)
    expect(context.setSearchResponse).toHaveBeenCalledWith('METAL')
  })
})
