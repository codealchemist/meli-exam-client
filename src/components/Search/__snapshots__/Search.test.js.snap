// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should render 1`] = `
<Search>
  <Consumer>
    <styled.div>
      <StyledComponent
        forwardedComponent={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "attrs": Array [],
            "componentStyle": ComponentStyle {
              "componentId": "sc-bdVaJa",
              "isStatic": true,
              "lastClassName": "gkztzw",
              "rules": Array [
                "
  flex: 1;
  position: relative;

  input {
    height: 39px;
    box-shadow: 0 1px 2px 0 rgba(0, 0, 0, 0.2);
    padding: 7px 60px 9px 15px;
    border-radius: 2px;
    background-color: ",
                "#fff",
                ";
    border: 0 rgba(0, 0, 0, 0.2);
    color: ",
                "#333",
                ";
    font-size: 16px;
    width: 100%;
    margin: 0;
  }
",
              ],
            },
            "displayName": "styled.div",
            "foldedComponentIds": Array [],
            "render": [Function],
            "styledComponentId": "sc-bdVaJa",
            "target": "div",
            "toString": [Function],
            "warnTooManyClasses": [Function],
            "withComponent": [Function],
          }
        }
        forwardedRef={null}
      >
        <div
          className="sc-bdVaJa gkztzw"
        >
          <input
            autoFocus={true}
            maxLength="120"
            onChange={[Function]}
            onKeyPress={[Function]}
            placeholder="Buscar productos, marcas y mÃ¡s..."
            type="search"
          />
          <div
            background="#eee"
            onClick={[Function]}
            type="nav-icon-search"
          />
        </div>
      </StyledComponent>
    </styled.div>
  </Consumer>
</Search>
`;
